BufferedRowResultProxy should stop growing at 100

mcclurem opened the pull request:
The class comment indicates we stop growing at 100 but was going to 1000. In "yield_per" queries that return large columns (in my case huge hstore columns) the memory usage can grow larger than anticipated and means that if you need to reduce your memory requirements lower than 1000 fetched rows you can't.

zzzeek commented:
OK, we can make that number change, but not like this.   An arbitrary number there that works for you means that it's arbitary; it won't work for someone else.  Hence the "top" number should be configurable.   This PR needs these modifications: 

1. the table of growth stays the same for now
2. the "top" number it can go to is available via a new value `self.context.execution_options.get('_max_row_buffer')`, if present
3. yield_per() on Query now also adds "max_row_buffer = <num>" to the execution options dictionary it sets up
4. this value needs a test in both test/engine/test_execute() with a Core execution as well as a query test in test/orm/test_query -> YieldTest.   If you need help with the tests then just give me 1-3 and I'll see what I can do.

mcclurem commented:
Fair enough!
I reverted the table to how it was, tweaked the comment to reflect the 1000 row max, and implemented 
the max_row_buffer behavior.
I built it and tested it's functional and has the desired impact both in terms of functional behavior and memory consumption.
Not certain about writing the tests, if you can bang them out easily that might be better, but I can take a look

zzzeek commented:
OK, this is enough for me to work with and I added tests in 9ccdea3a0fe57931e779b44e.  thanks for contributing!

dmerrick commented:
This is the greatest PR in the history of man.
